### PR TITLE
Improve project parallel loading with credentials requested

### PR DESCRIPTION
**Pull request by vcloarec**
This PR improves parallel loading (see #53069 ) when there are some credentials requested. 
Before this PR, the parallel loading skip all layers that require credentials.

With this PR, when a attempt returns a invalid layer, as it could be a layer that request credentials to be entered by user, the parallel loading stop, try to open the layer in the main thread, so credentials are eventually requested and are stored  in the cache, then restart parallel loading. If there are some layers that request the same credential, there are in the cache and the layers are loaded in parallel.

### CODE REVIEW

File: src/core/project/qgsproject.cpp [hunk 1]
```diff
@@ -1324,63 +1324,117 @@ static  QgsMapLayer::ReadFlags projectFlagsToLayerReadFlags( Qgis::ProjectReadFl
...
+
+    threadPool.waitForDone(); // to be sure all threads are finished
+
+    // First we take all valid providers
+    for ( QgsRunnableProviderCreator *run : std::as_const( runnables ) )
+    {
+      std::unique_ptr<QgsDataProvider> provider( run->dataProvider() );
+      if ( provider && provider->isValid() )
+        loadedProviders.insert( run->layerId(), provider.release() );
+    }
```
Comment (wonder-sk) [inline src/core/project/qgsproject.cpp:1413]: Can we insert to the `loadedProviders` directly in the slot for `QgsRunnableProviderCreator::providerCreated` when `isValid` is true? That would make the code a bit easier to understand, as this for-loop would not be needed...?

File: src/core/project/qgsproject.cpp [hunk 2]
```diff
@@ -1324,63 +1324,117 @@ static  QgsMapLayer::ReadFlags projectFlagsToLayerReadFlags( Qgis::ProjectReadFl
...
+          threadPool.clear(); //we have to stop all loading provider to try this layer in main thread and maybe have credentials
+        }
+
+        if ( i == parallelLayerNodes.count() || !isValid )
+          loop.quit();
+      } );
+      threadPool.start( run );
+    }
+    if ( !parallelLayerNodes.isEmpty() )
+      loop.exec();
```
Comment (wonder-sk) [inline src/core/project/qgsproject.cpp:1403]: Isn't `parallelLayerNodes` always going to be non-empty when we're in this while loop? can we get rid of the if, and always run loop.exec() ?

**Review by wonder-sk — COMMENTED** [review]
I have tested this and it seems to work fine - I get credentials dialog when needed and then parallel loading continues... good work!

File: src/core/project/qgsproject.cpp [hunk 3]
```diff
@@ -1324,63 +1324,117 @@ static  QgsMapLayer::ReadFlags projectFlagsToLayerReadFlags( Qgis::ProjectReadFl
...
+    // We try with the first layer returned invalid but this time in the main thread to maybe have credentials and continue with others not loaded in parallel
+    QMap<QString, LayerToLoad>::ConstIterator it = layersToLoad.find( layerToAttemptInMainThread );
+    if ( it != layersToLoad.constEnd() )
+    {
+      const LayerToLoad &lay =  it.value();
+      QgsDataProvider::ReadFlags providerFlags = lay.flags;
+      providerFlags.setFlag( QgsDataProvider::SkipCredentialsRequest, false );
+      QgsScopedRuntimeProfile profile( "Create data providers/" + lay.layerId, QStringLiteral( "projectload" ) );
+      std::unique_ptr<QgsDataProvider> provider( QgsProviderRegistry::instance()->createProvider( lay.provider, lay.dataSource, lay.options, providerFlags ) );
       i++;
```
[the remark this answers is not in the document]
Comment (vcloarec) [reply src/core/project/qgsproject.cpp:1426]: all provider, valid and not valid should be counted. Here in the main thread, if the layer is invalid, there will not have another attempt.

File: src/core/project/qgsproject.cpp (hunk 2 above)
Comment (vcloarec) [reply src/core/project/qgsproject.cpp:1403]: I guess it is a leftover...

File: src/core/project/qgsproject.cpp (hunk 1 above)
Comment (vcloarec) [reply src/core/project/qgsproject.cpp:1413]: let's try.

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**